dial_queue.go

```diff
@@ -91,18 +91,23 @@ func (dq *dialQueue) control() {
 		waiting        <-chan waitingCh
 		lastScalingEvt = time.Now()
 	)
+
+	defer func() {
+		// close channels.
+		if resp.ch != nil {
+			close(resp.ch)
+		}
+		close(dq.waitingCh)
```

Comment:
Is this safe? Aren't we going to get a concurrent `Consume()`?

I think the solution here *may* be to:

1. Make this channel synchronous.
2. Buffer internally (in a slice).

I can't think of a non-racy alternative.

Reply:
I thought a select with a default branch, where one of the cases was sending on a closed channel, wouldn't panic. But it does. 

Would setting the `dq.waitingCh = nil` before closing would remove the race?

Regardless, I'm starting to think the buffering in a slice is a good option.

Reply:
I think the only other way is to lock around writes to `waitingCh` but that gets ugly fast (see the entire go-ipfs commands library).